shared/src/econ/mod.rs

```diff
@@ -1,7 +1,345 @@
-// SPDX-License-Identifier: Apache-2.0, MIT
+use std::cmp::Ordering;
+use std::fmt;
+use std::ops::{Add, Mul, Neg, Sub};
 
 use num_bigint::BigInt;
+use num_integer::Integer;
+use num_traits::{Signed, Zero};
+use serde::{Deserialize, Serialize, Serializer};
 
-/// Token type to be used within the VM.
-pub type TokenAmount = BigInt;
+use crate::bigint::bigint_ser;
+
+/// A quantity of native tokens.
+/// A token amount is an integer, but has a human interpretation as a value with
+/// 18 decimal places.
+/// This is a new-type in order to prevent accidental conversion from other BigInts.
+/// From/Into BigInt is missing by design.
+#[derive(Clone, PartialEq, Eq, Hash)]
+pub struct TokenAmount {
+    atto: BigInt,
+}
+
+// This type doesn't implement all the numeric traits (Num, Signed, etc),
+// opting for a minimal useful set. Others can be added if needed.
+impl TokenAmount {
+    /// The logical number of decimal places of a token unit.
+    pub const DECIMALS: usize = 18;
+
+    /// The logical precision of a token unit.
+    pub const PRECISION: u64 = 10u64.pow(Self::DECIMALS as u32);
+
+    /// Creates a token amount from a quantity of indivisible units  (10^-18 whole units).
+    pub fn from_atto(atto: BigInt) -> Self {
+        Self { atto }
+    }
+
+    /// Creates a token amount from a quantity of indivisible units (10^-18 whole units).
+    pub fn from_atto_u128(atto: u128) -> Self {
+        Self {
+            atto: BigInt::from(atto),
+        }
+    }
+
+    /// Creates a token amount from a quantity of indivisible units (10^-18 whole units).
+    pub fn from_atto_u64(atto: u64) -> Self {
+        Self {
+            atto: BigInt::from(atto),
+        }
+    }
+
+    /// Creates a token amount from a quantity of whole units (10^18 indivisible units).
+    pub fn from_whole(tokens: i64) -> Self {
+        Self {
+            atto: BigInt::from(tokens) * Self::PRECISION,
+        }
+    }
+
+    pub fn zero() -> Self {
+        Self {
+            atto: BigInt::zero(),
+        }
+    }
+
+    /// Returns the quantity of indivisible units.
+    pub fn atto(&self) -> &BigInt {
+        &self.atto
+    }
+
+    pub fn is_zero(&self) -> bool {
+        self.atto.is_zero()
+    }
+
+    pub fn is_positive(&self) -> bool {
+        self.atto.is_positive()
+    }
+
+    pub fn is_negative(&self) -> bool {
+        self.atto.is_negative()
+    }
+}
+
+impl PartialOrd for TokenAmount {
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        self.atto.partial_cmp(&other.atto)
+    }
+}
+
+impl Ord for TokenAmount {
+    #[inline]
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.atto.cmp(&other.atto)
+    }
+}
+
+impl Default for TokenAmount {
+    #[inline]
+    fn default() -> TokenAmount {
+        TokenAmount::zero()
+    }
+}
+
+impl fmt::Debug for TokenAmount {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(self, f)
```

Comment:
It might be helpful to keep the newtype here (pretty common in debug impls.

```suggestion
        write!(f, "TokenAmount({})", self)
```

But I don't feel strongly about this.